log/logger_impl.go

```diff
@@ -0,0 +1,132 @@
+package log
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+)
+
+func NewDefaultLogger() Logger {
+	return NewLogger(log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr)))
+}
+
+func NewNopLogger() Logger {
+	return NewLogger(log.NewNopLogger())
+}
+
+func NewJSONLogger() Logger {
+	return NewLogger(log.NewJSONLogger(log.NewSyncWriter(os.Stderr)))
+}
+
+func NewBufferLogger() (*strings.Builder, Logger) {
+	buffer := strings.Builder{}
+	writer := log.NewLogfmtLogger(log.NewSyncWriter(&buffer))
+	log := NewLogger(writer)
+	return &buffer, log
+}
+
+func NewLogger(writer log.Logger) Logger {
+	l := &logger{
+		writer:  writer,
+		ctx:     map[string]string{},
+		lastErr: nil,
+	}
+
+	// Default logs to be info until changed
+	return l.Info()
+}
+
+type logger struct {
+	writer  log.Logger
+	ctx     map[string]string
+	lastErr error
+}
+
+func (l *logger) Set(key, value string) Logger {
+	return l.With(Fields{
+		key: value,
+	})
+}
+
+// With returns a new Logger with the contexts added to its own.
+func (l *logger) With(ctxs ...Context) Logger {
+	// Estimation assuming that for each ctxs has at least 1 value.
+	combined := make(map[string]string, len(l.ctx)+len(ctxs))
+
+	for k, v := range l.ctx {
+		combined[k] = v
+	}
+
+	for _, c := range ctxs {
+		itemCtx := c.Context()
+		for k, v := range itemCtx {
+			combined[k] = v
+		}
+	}
+
+	return &logger{
```

Comment:
We should reuse our instance of a `logger`.

Here we can do
```go
l.combined = combined
return l
```

Reply:
You don't want to re-use it because the mutability would affect other instances of it and make it not thread safe anymore.

Reply:
Ah I see.